File: api/src/main/java/dev/obscuria/fragmentum/ModCompat.java
package dev.obscuria.fragmentum;

import org.apache.commons.lang3.NotImplementedException;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.function.Supplier;

@SuppressWarnings("unused")
public record ModCompat(String modId, String modName) {

    public ModCompat(String modId) {
        this(modId, modId);
    }

    public String ModName() {
        throw new NotImplementedException();
    }

    public boolean isLoaded() {
        throw new NotImplementedException();
    }

    public <T> @Nullable T getIfLoaded(Supplier<Supplier<T>> supplier) {
        throw new NotImplementedException();
    }

    public void runIfLoaded(Supplier<Runnable> runnable) {
        throw new NotImplementedException();
    }

    public void runIfMissing(Runnable runnable) {
        throw new NotImplementedException();
    }
}

File: api/src/main/kotlin/dev/obscuria/fragmentum/ModCompat.kt
@file:Suppress("unused", "UNUSED_PARAMETER", "RedundantNullableReturnType")

package dev.obscuria.fragmentum

class ModCompat {

    constructor(modId: String)

    constructor(modId: String, displayName: String)

    val modId: String get() = error("API stub")

    val displayName: String get() = error("API stub")

    val isLoaded: Boolean get() = error("API stub")

    inline fun <T> getIfLoaded(supplier: () -> () -> T): T? = error("API stub")

    inline fun runIfLoaded(runnable: () -> () -> Any): Unit = error("API stub")

    inline fun runIfMissing(runnable: () -> Any): Unit = error("API stub")
}